### PR TITLE
Axi2csr srams

### PR DESCRIPTION
**Pull request by Spaqin**
Previous pull request I made - native AXI interface for SRAM - turned out to be insufficient. If you needed more than one memory space, you were out of luck - only two AXI slaves are currently supported. And yet it was necessary for kasli-soc drtio support, as it requires multiple memory-mapped memories for its aux controllers.

Instead on the suggestion of @sbourdeauducq I went with a different approach, in this PR - used the existing axi2csr bridge and added SRAM support to it, extending the bus. Original CSR address space is untouched, but the bridge has an internal bus with address decoder that connect together the CSR bus with any additional memories or slaves that have the same interface as CSR. This of course requires an extension of the address bus which is configurable because it depends of course on how much more memory you want to add on top of that. It may seem somewhat wrong as it's still called an axi2csr bridge, but integrating them together allows you to make use of one axi slave for multiple purposes, allowing multiple memories to work within the limitations.

Since this allows you to add ports, whether they're rw or read only, or just memory (either created beforehand, or by passing size, like in most previous SRAM implementations), included tests (based again on axi2csr :)) should cover all scenarios. I'm open to any suggestions and improvements, maybe some variable names aren't clear.

### CODE REVIEW

File: src/migen_axi/integration/soc_core.py [hunk 1]
```diff
@@ -54,11 +55,16 @@ def __init__(self, platform,
             ddr=platform.request("ddr"),
         ))
 
+        # csr_addr_extension is by how many address lanes over csr
+        # should be widened to accommodate for any additional
+        # memories in axi2csr
```
Comment (sbourdeauducq) [inline src/migen_axi/integration/soc_core.py:60]: Can't it be determined automatically based on the memory ports that have been added?

File: src/migen_axi/interconnect/axi2csr.py [hunk 2]
```diff
@@ -87,7 +102,104 @@ def __init__(self, bus_axi=None, bus_csr=None):
...
+            self._internal_csr.we.eq(0),
+            self._internal_csr.dat_w.eq(w.data),
         ]
+
+    def add_slave(self, fun, slave):
+        # fun - function that takes the adr signal and returns a FHDL expr.
+        #       that evaluates to 1 when the slave is selected and 0 otherwise.
+        # slave - Memory.port or csr bus reference.
+        if isinstance(slave, fhdl.specials.Special):
+            self.specials += slave
```
Comment (sbourdeauducq) [inline src/migen_axi/interconnect/axi2csr.py:115]: Leave that to the caller.

Comment (Spaqin) [reply src/migen_axi/interconnect/axi2csr.py]: Makes sense, done.

File: src/migen_axi/interconnect/axi2csr.py [hunk 3]
```diff
@@ -7,19 +7,18 @@
...
         self.csr = bus_csr or csr_bus.Interface()
 
-        # internal bus that will connect to address decoder
-        # needs to be wider than CSR bus to accommodate both bus and any SRAM
-        # data width also identical to AXI to accomodate SRAM
-        self._internal_csr = csr_bus.Interface(
-            data_width=len(self.bus.r.data),
-            address_width=len(self.csr.adr) + addr_extension)
+        # if None, will be calculated automatically
+        # if a number, it will be added to csr bus width
```
Comment (sbourdeauducq) [inline src/migen_axi/interconnect/axi2csr.py:15]: I would remove support for putting a fixed number entirely.

Comment (Spaqin) [reply src/migen_axi/interconnect/axi2csr.py]: ok done, in this case add_slave was also made a protected method (as in, name changed to _add_slave), so the chance of meddling with address space decreases further

**Comment by sbourdeauducq** [comment]
Looks acceptable to me, @peteut could you take a look and merge?

**Comment by peteut** [comment]
@Spaqin, @sbourdeauducq thanks, that looks fine.

**Pull request merged**